File: Examples/intel_examples/examples_openmp/f_openmp/vml/vlinearfrac.f90
!===============================================================================
! Copyright 2020 Intel Corporation.
!
! This software and the related documents are Intel copyrighted  materials,  and
! your use of  them is  governed by the  express license  under which  they were
! provided to you (License).  Unless the License provides otherwise, you may not
! use, modify, copy, publish, distribute,  disclose or transmit this software or
! the related documents without Intel's prior written permission.
!
! This software and the related documents  are provided as  is,  with no express
! or implied  warranties,  other  than those  that are  expressly stated  in the
! License.
!===============================================================================

!*
!
!*  Content:
!*            LinearFrac example program text (OpenMP offload interface)
!*
!*******************************************************************************/

include "mkl_omp_offload.f90"
include "_vml_common_functions.f90"

! @brief Real single precision function test begin
integer (kind=4) function test_float(funcname)
    use onemkl_vml_omp_offload
    implicit none
    include "_vml_common_data.f90"
    character (len = *) :: funcname
    real      (kind=4)  :: as_float
    integer   (kind=4)  :: check_result_float
    real      (kind=4),allocatable :: varg1(:), varg2(:), vres1(:), vmres1(:), vref1(:)
    real      (kind=4) arg3, arg4, arg5, arg6
    integer   (kind=4) i, a, errs
    integer   (kind=4) VLEN
    parameter (VLEN = 4)
    integer   (kind=4) test_arg1(VLEN)
    integer   (kind=4) test_arg2(VLEN)
    integer   (kind=4) test_arg3(VLEN)
    integer   (kind=4) test_arg4(VLEN)
    integer   (kind=4) test_arg5(VLEN)
    integer   (kind=4) test_arg6(VLEN)
    integer   (kind=4) test_ref1(VLEN)
    integer   (kind=4) nan_value
    integer   (kind=8) vml_accuracy_mode(3)
    data vml_accuracy_mode / VML_HA, VML_LA, VML_EP /
    integer tmode

    ! NaN value to fill result vector
    data  nan_value /Z'FFFFFFFF'/
    
    ! Arguments and reference results begin
    data test_arg1 / Z'40D9B85C', & ! 6.80375481     
                     Z'40B52EFA', & ! 5.66198444     
                     Z'4103BA28', & ! 8.2329483      
                     Z'C052EA36'  / ! -3.2955451     
    data test_arg2 / Z'C007309A', & ! -2.1123414     
                     Z'40BF006A', & ! 5.96880054     
                     Z'C0C1912F', & ! -6.04897261    
                     Z'40ABAABC'  / ! 5.3645916      
    data test_arg3 / Z'4048F5C3', & ! 3.1400001      
                     Z'4048F5C3', & ! 3.1400001      
                     Z'4048F5C3', & ! 3.1400001      
                     Z'4048F5C3'  / ! 3.1400001      
    data test_arg4 / Z'40C8F5C3', & ! 6.28000021     
                     Z'40C8F5C3', & ! 6.28000021     
                     Z'40C8F5C3', & ! 6.28000021     
                     Z'40C8F5C3'  / ! 6.28000021     
    data test_arg5 / Z'4116B852', & ! 9.42000008     
                     Z'4116B852', & ! 9.42000008     
                     Z'4116B852', & ! 9.42000008     
                     Z'4116B852'  / ! 9.42000008     
    data test_arg6 / Z'4148F5C3', & ! 12.5600004     
                     Z'4148F5C3', & ! 12.5600004     
                     Z'4148F5C3', & ! 12.5600004     
                     Z'4148F5C3'  / ! 12.5600004     
    data test_ref1 / Z'C07117D3', & ! -3.76707911    
                     Z'3EB313C1', & ! 0.349760085    
                     Z'BF392C6C', & ! -0.723334074   
                     Z'BD840B70'  / ! -0.0644749403  
    ! Arguments and reference results end
    
    errs = 0

    ! Allocate vectors
    allocate(varg1(VLEN))
    allocate(varg2(VLEN))
    allocate(vres1(VLEN))
    allocate(vmres1(VLEN))
    allocate(vref1(VLEN))

    ! Fill vectors
    do i = 1, VLEN
        varg1(i) = as_float(test_arg1(i))
        varg2(i) = as_float(test_arg2(i))
        arg3     = as_float(test_arg3(i))
        arg4     = as_float(test_arg4(i))
        arg5     = as_float(test_arg5(i))
        arg6     = as_float(test_arg6(i))
        vref1(i) = as_float(test_ref1(i))
        vres1(i) = as_float(nan_value)
        vmres1(i) = as_float(nan_value)
    enddo

    ! Loop by three accuracy flavors
    do a = 1, 3
        ! Call VML function with specific accuracy flavor
        !$omp target variant dispatch
        tmode = vmlsetmode(vml_accuracy_mode(a))
        !$omp end target variant dispatch

        !$omp target data map(varg1,varg2,vres1)
        !$omp target variant dispatch use_device_ptr(varg1,varg2,vres1)
        call vslinearfrac(VLEN, varg1, varg2, arg3, arg4, arg5, arg6, vres1)
        !$omp end target variant dispatch
        !$omp end target data
        !$omp target data map(varg1,varg2,vmres1)
        !$omp target variant dispatch use_device_ptr(varg1,varg2,vmres1)
        call vmslinearfrac(VLEN, varg1, varg2, arg3, arg4, arg5, arg6, vmres1, vml_accuracy_mode(a))
        !$omp end target variant dispatch
        !$omp end target data
        ! Check results
        do i = 1, VLEN
          errs = errs + check_result_float(VML_ARG1_RES1, varg1(i), varg2(i), & 
                           vres1(i), vres1(i), vref1(i), vref1(i), "v"//funcname, a)
          errs = errs + check_result_float(VML_ARG1_RES1, varg1(i), varg2(i), & 
                           vmres1(i), vmres1(i), vref1(i), vref1(i), "vm"//funcname, a)
        enddo
    enddo

    test_float = errs

end function
! @brief Real single precision function test end

! @brief Real double precision function test begin
integer (kind=4) function test_double(funcname)
    use onemkl_vml_omp_offload
    implicit none
    include "_vml_common_data.f90"
    character (len = *) :: funcname
    real      (kind=8) :: as_double
    integer   (kind=4) :: check_result_double
    real      (kind=8),allocatable :: varg1(:), varg2(:), vres1(:), vmres1(:), vref1(:)
    real      (kind=8) arg3, arg4, arg5, arg6
    integer   (kind=4) i, a, errs
    integer   (kind=8) VLEN
    parameter (VLEN = 4)
    integer   (kind=8) test_arg1(VLEN)
    integer   (kind=8) test_arg2(VLEN)
    integer   (kind=8) test_arg3(VLEN)
    integer   (kind=8) test_arg4(VLEN)
    integer   (kind=8) test_arg5(VLEN)
    integer   (kind=8) test_arg6(VLEN)
    integer   (kind=8) test_ref1(VLEN)
    integer   (kind=8) nan_value
    integer   (kind=8) vml_accuracy_mode(3)
    data vml_accuracy_mode / VML_HA, VML_LA, VML_EP /
    integer tmode

    ! NaN value to fill result vector
    data  nan_value /Z'FFFFFFFFFFFFFFFF'/
    
    ! Arguments and reference results begin
    data test_arg1 / Z'401B370B60E66E18', & ! 6.80375434309419092      
                     Z'4016A5DF421D4BBE', & ! 5.66198447517211711      
                     Z'40207744D998EE8A', & ! 8.23294715873568705      
                     Z'C00A5D46A314BA8E'  / ! -3.2955448857022196      
    data test_arg2 / Z'C000E6134801CC26', & ! -2.11234146361813924     
                     Z'4017E00D485FC01A', & ! 5.96880066952146571      
                     Z'C0183225E080644C', & ! -6.04897261413232101     
                     Z'4015755793FAEAB0'  / ! 5.36459189623808186      
    data test_arg3 / Z'40091EB851EB851F', & ! 3.14000000000000012      
                     Z'40091EB851EB851F', & ! 3.14000000000000012      
                     Z'40091EB851EB851F', & ! 3.14000000000000012      
                     Z'40091EB851EB851F'  / ! 3.14000000000000012      
    data test_arg4 / Z'40191EB851EB851F', & ! 6.28000000000000025      
                     Z'40191EB851EB851F', & ! 6.28000000000000025      
                     Z'40191EB851EB851F', & ! 6.28000000000000025      
                     Z'40191EB851EB851F'  / ! 6.28000000000000025      
    data test_arg5 / Z'4022D70A3D70A3D7', & ! 9.41999999999999993      
                     Z'4022D70A3D70A3D7', & ! 9.41999999999999993      
                     Z'4022D70A3D70A3D7', & ! 9.41999999999999993      
                     Z'4022D70A3D70A3D7'  / ! 9.41999999999999993      
    data test_arg6 / Z'40291EB851EB851F', & ! 12.5600000000000005      
                     Z'40291EB851EB851F', & ! 12.5600000000000005      
                     Z'40291EB851EB851F', & ! 12.5600000000000005      
                     Z'40291EB851EB851F'  / ! 12.5600000000000005      
    data test_ref1 / Z'C00E22FA0DD4CC16', & ! -3.76707850270896483     
                     Z'3FD6627804818A52', & ! 0.349760059738547402     
                     Z'BFE7258D6AC8E3FB', & ! -0.723334034503863577    
                     Z'BFB0816DEA262188'  / ! -0.0644749352123935582   
    ! Arguments and reference results end
    
    errs = 0

    ! Allocate vectors
    allocate(varg1(VLEN))
    allocate(varg2(VLEN))
    allocate(vres1(VLEN))
    allocate(vmres1(VLEN))
    allocate(vref1(VLEN))

    ! Fill vectors
    do i = 1, VLEN
        varg1(i) = as_double(test_arg1(i))
        varg2(i) = as_double(test_arg2(i))
        arg3     = as_double(test_arg3(i))
        arg4     = as_double(test_arg4(i))
        arg5     = as_double(test_arg5(i))
        arg6     = as_double(test_arg6(i))
        vref1(i) = as_double(test_ref1(i))
        vres1(i) = as_double(nan_value)
        vmres1(i) = as_double(nan_value)
    enddo

    ! Loop by three accuracy flavors
    do a = 1, 3
        ! Call VML function with specific accuracy flavor
        !$omp target variant dispatch
        tmode = vmlsetmode(vml_accuracy_mode(a))
        !$omp end target variant dispatch

        !$omp target data map(varg1,varg2,vres1)
        !$omp target variant dispatch use_device_ptr(varg1,varg2,vres1)
        call vdlinearfrac(VLEN, varg1, varg2, arg3, arg4, arg5, arg6, vres1)
        !$omp end target variant dispatch
        !$omp end target data
        !$omp target data map(varg1,varg2,vmres1)
        !$omp target variant dispatch use_device_ptr(varg1,varg2,vmres1)
        call vmdlinearfrac(VLEN, varg1, varg2, arg3, arg4, arg5, arg6, vmres1, vml_accuracy_mode(a))
        !$omp end target variant dispatch
        !$omp end target data
        ! Check results
        do i = 1, VLEN
          errs = errs + check_result_double(VML_ARG1_RES1, varg1(i), varg2(i), & 
                            vres1(i), vres1(i), vref1(i), vref1(i), "v"//funcname, a)
          errs = errs + check_result_double(VML_ARG1_RES1, varg1(i), varg2(i), & 
                            vmres1(i), vmres1(i), vref1(i), vref1(i), "vm"//funcname, a)
        enddo
    enddo
    
    test_double = errs

end function
! @brief Real double precision function test end

! @brief Main test program begin
program linearfrac_example
    use onemkl_vml_omp_offload
    implicit none
    include "_vml_common_data.f90"
    integer   (kind=4) :: blend_int32
    integer   (kind=4) :: test_float
    integer   (kind=4) :: test_double
    integer   (kind=4) errs, total_errs
    character (len = *), parameter :: funcname = "linearfrac"
    
    total_errs = 0

    data FLOAT_MAXULP /1.0D6,1.0D6,1.0D6/
    data COMPLEX_FLOAT_MAXULP /FLOAT_COMPLEX_MAXULP_HA,FLOAT_COMPLEX_MAXULP_LA,FLOAT_COMPLEX_MAXULP_EP/
    data DOUBLE_MAXULP /1.0D6,1.0D6,1.0D6/
    data COMPLEX_DOUBLE_MAXULP /DOUBLE_COMPLEX_MAXULP_HA,DOUBLE_COMPLEX_MAXULP_LA,DOUBLE_COMPLEX_MAXULP_EP/

    write (*, 111) funcname
    111 format ('Running ', A, ' functions:')

    ! Single precision test run begin
    write (*, 112) TAB, funcname
    112 format(A, 'Running ',  A, ' with single precision real data type:')
    errs = test_float(funcname)    
    write (*, 113) TAB, funcname, TEST_RESULT(blend_int32((errs>0),2,1))
    113 format(A, A, ' single precision real result: ', A)
    ! Single precision test run end

    ! Real double precision test run begin
    write (*, 117) TAB, funcname
    117 format(A, 'Running ',  A, ' with double precision real data type:')
    errs = test_double(funcname)    
    write (*, 118) TAB, funcname, TEST_RESULT(blend_int32((errs>0),2,1))
    118 format(A, A, ' double precision real result: ', A)
    ! Real double precision test run end

    write (*, 121) funcname, TEST_RESULT(blend_int32((total_errs>0),2,1))
    121 format(A, ' function result: ', A)

end program
! @brief Main test program end

File: Examples/intel_examples/examples_openmp/c_openmp/vml/vlinearfrac.c
/*******************************************************************************
* Copyright 2018-2020 Intel Corporation.
*
* This software and the related documents are Intel copyrighted  materials,  and
* your use of  them is  governed by the  express license  under which  they were
* provided to you (License).  Unless the License provides otherwise, you may not
* use, modify, copy, publish, distribute,  disclose or transmit this software or
* the related documents without Intel's prior written permission.
*
* This software and the related documents  are provided as  is,  with no express
* or implied  warranties,  other  than those  that are  expressly stated  in the
* License.
*******************************************************************************/

/*
 *
 *  Content:
 *            LinearFrac example program text (OpenMP offload interface)
 *
 *******************************************************************************/

#define VLEN 4

#include "_vml_common.h"

max_ulp_table_t max_ulp_table =
{ //  HA   LA   EP
    { 1.0E6, 1.0E6, 1.0E6, }, // float
    { 1.0E6, 1.0E6, 1.0E6, }, // double
    { 2.0, 5.0, 5.0E3, }, // VM_COMPLEX8
    { 2.0, 5.0, 7.0E7, }, // VM_COMPLEX16
};

// device number
int dnum;

// *************************************************************
// Data table declaraion
// *************************************************************
data_7_t data =
{

{

{ { 0x40D9B85C }, { 0xC007309A }, { 0x4048F5C3 }, { 0x40C8F5C3 }, { 0x4116B852 }, { 0x4148F5C3 }, { 0xC07117D3 } }, //  0: vsLinearFrac ( 6.80375481     , -2.1123414     , 3.1400001      , 6.28000021     , 9.42000008     , 12.5600004      ) = ( -3.76707911     );
{ { 0x40B52EFA }, { 0x40BF006A }, { 0x4048F5C3 }, { 0x40C8F5C3 }, { 0x4116B852 }, { 0x4148F5C3 }, { 0x3EB313C1 } }, //  1: vsLinearFrac ( 5.66198444     , 5.96880054     , 3.1400001      , 6.28000021     , 9.42000008     , 12.5600004      ) = ( 0.349760085     );
{ { 0x4103BA28 }, { 0xC0C1912F }, { 0x4048F5C3 }, { 0x40C8F5C3 }, { 0x4116B852 }, { 0x4148F5C3 }, { 0xBF392C6C } }, //  2: vsLinearFrac ( 8.2329483      , -6.04897261    , 3.1400001      , 6.28000021     , 9.42000008     , 12.5600004      ) = ( -0.723334074    );
{ { 0xC052EA36 }, { 0x40ABAABC }, { 0x4048F5C3 }, { 0x40C8F5C3 }, { 0x4116B852 }, { 0x4148F5C3 }, { 0xBD840B70 } }, //  3: vsLinearFrac ( -3.2955451     , 5.3645916      , 3.1400001      , 6.28000021     , 9.42000008     , 12.5600004      ) = ( -0.0644749403   );
}

,

{

{ { 0x401B370B60E66E18 }, { 0xC000E6134801CC26 }, { 0x40091EB851EB851F }, { 0x40191EB851EB851F }, { 0x4022D70A3D70A3D7 }, { 0x40291EB851EB851F }, { 0xC00E22FA0DD4CC16 } }, //  0: vdLinearFrac ( 6.80375434309419092      , -2.11234146361813924     , 3.14000000000000012      , 6.28000000000000025      , 9.41999999999999993      , 12.5600000000000005       ) = ( -3.76707850270896483      );
{ { 0x4016A5DF421D4BBE }, { 0x4017E00D485FC01A }, { 0x40091EB851EB851F }, { 0x40191EB851EB851F }, { 0x4022D70A3D70A3D7 }, { 0x40291EB851EB851F }, { 0x3FD6627804818A52 } }, //  1: vdLinearFrac ( 5.66198447517211711      , 5.96880066952146571      , 3.14000000000000012      , 6.28000000000000025      , 9.41999999999999993      , 12.5600000000000005       ) = ( 0.349760059738547402      );
{ { 0x40207744D998EE8A }, { 0xC0183225E080644C }, { 0x40091EB851EB851F }, { 0x40191EB851EB851F }, { 0x4022D70A3D70A3D7 }, { 0x40291EB851EB851F }, { 0xBFE7258D6AC8E3FB } }, //  2: vdLinearFrac ( 8.23294715873568705      , -6.04897261413232101     , 3.14000000000000012      , 6.28000000000000025      , 9.41999999999999993      , 12.5600000000000005       ) = ( -0.723334034503863577     );
{ { 0xC00A5D46A314BA8E }, { 0x4015755793FAEAB0 }, { 0x40091EB851EB851F }, { 0x40191EB851EB851F }, { 0x4022D70A3D70A3D7 }, { 0x40291EB851EB851F }, { 0xBFB0816DEA262188 } }, //  3: vdLinearFrac ( -3.2955448857022196      , 5.36459189623808186      , 3.14000000000000012      , 6.28000000000000025      , 9.41999999999999993      , 12.5600000000000005       ) = ( -0.0644749352123935582    );
}

,
};

//!
//! @brief Single precision test
//!

int vLinearFracAccuracyLiteTest_float() {
  int errs = 0;
  float * varg1 = (float *)malloc(VLEN * sizeof(float));

  float * varg2 = (float *)malloc(VLEN * sizeof(float));

  float varg3;
  float varg4;
  float varg5;
  float varg6;

  float * vres1 = (float *)malloc(VLEN * sizeof(float));
  float * vref1 = (float *)malloc(VLEN * sizeof(float));

  {
    for (int i = 0; i < VLEN; ++i) {
      varg1[i] = data.data_f32[i].v1.f;
      varg2[i] = data.data_f32[i].v2.f;
      varg3 = data.data_f32[i].v3.f;
      varg4 = data.data_f32[i].v4.f;
      varg5 = data.data_f32[i].v5.f;
      varg6 = data.data_f32[i].v6.f;
      vref1[i] = data.data_f32[i].v7.f;
    }
  }

  for (int acc = 0; acc < ACCURACY_NUM; ++acc) {
#pragma omp target data map(to:varg1[0:VLEN]) map(to:varg2[0:VLEN]) map(tofrom:vres1[0:VLEN]) device(dnum)
    {
#pragma omp target variant dispatch device(dnum) use_device_ptr(varg1, varg2, vres1) nowait
      {
        vmsLinearFrac(VLEN, (const float *)varg1, (const float *)varg2,
                      (const float)varg3, (const float)varg4,
                      (const float)varg5, (const float)varg6, (float *)vres1,
                      accuracy_mode[acc]);
      }
#pragma omp taskwait
    }

    for (int i = 0; i < VLEN; ++i) {

      errs +=
          check_result_float(ARG2_RES1, varg1[i], varg2[i], vres1[i], vres1[i],
                             vref1[i], vref1[i], "LinearFrac", acc);
    }
  }

  free(varg1);

  free(varg2);

  free(vres1);
  free(vref1);

  return errs;
}
//!
//! @brief Double precision test
//!

int vLinearFracAccuracyLiteTest_double() {
  int errs = 0;
  double * varg1 = (double *)malloc(VLEN * sizeof(double));

  double * varg2 = (double *)malloc(VLEN * sizeof(double));

  double varg3;
  double varg4;
  double varg5;
  double varg6;

  double * vres1 = (double *)malloc(VLEN * sizeof(double));
  double * vref1 = (double *)malloc(VLEN * sizeof(double));

  {
    for (int i = 0; i < VLEN; ++i) {
      varg1[i] = data.data_f64[i].v1.f;
      varg2[i] = data.data_f64[i].v2.f;
      varg3 = data.data_f64[i].v3.f;
      varg4 = data.data_f64[i].v4.f;
      varg5 = data.data_f64[i].v5.f;
      varg6 = data.data_f64[i].v6.f;
      vref1[i] = data.data_f64[i].v7.f;
    }
  }

  for (int acc = 0; acc < ACCURACY_NUM; ++acc) {
#pragma omp target data map(to:varg1[0:VLEN]) map(to:varg2[0:VLEN]) map(tofrom:vres1[0:VLEN]) device(dnum)
    {
#pragma omp target variant dispatch device(dnum) use_device_ptr(varg1, varg2, vres1) nowait
      {
        vmdLinearFrac(VLEN, (const double *)varg1, (const double *)varg2,
                      (const double)varg3, (const double)varg4,
                      (const double)varg5, (const double)varg6, (double *)vres1,
                      accuracy_mode[acc]);
      }
#pragma omp taskwait
    }

    for (int i = 0; i < VLEN; ++i) {

      errs +=
          check_result_double(ARG2_RES1, varg1[i], varg2[i], vres1[i], vres1[i],
                              vref1[i], vref1[i], "LinearFrac", acc);
    }
  }

  free(varg1);

  free(varg2);

  free(vres1);
  free(vref1);

  return errs;
}

int main(int argc, char **argv) {
  int errs = 0;
  int total_errs = 0;

  printf("Running %s functions:\n", "LinearFrac");

  printf("\tRunning %s with single precision real data type:\n", "LinearFrac");
  errs = vLinearFracAccuracyLiteTest_float();
  printf("\t%s single precision real result: %s\n\n", "LinearFrac",
         (errs == 0) ? "PASS" : "FAIL");
  total_errs += errs;

  printf("\tRunning %s with double precision real data type:\n", "LinearFrac");
  errs = vLinearFracAccuracyLiteTest_double();
  printf("\t%s double precision real result: %s\n\n", "LinearFrac",
         (errs == 0) ? "PASS" : "FAIL");
  total_errs += errs;
  printf("%s function result: %s\n\n", "LinearFrac",
         (total_errs == 0) ? "PASS" : "FAIL");

  return total_errs;
}
